Draft: My memory-issue-playground

volkerwahh opened the pull request:
## Before Intro
Hi @pcorless,
Thx a lot for providing/maintaining icepdf! Would be nice, if u could have a look at this "local" PR. I did not references this pull request to the original [pcorless/icepdf](https://github.com/pcorless/icepdf) as this is more intended to be a base for discussion about what make sense and what not ( and it's my first PR at github and I am a shy backend-programmer ;)).  

## Intro

We use icepdf in our company in our internal client for user to view pdfs and select pages from the pdfs for further processing and we are pretty happy with it! From time to time we have some memory-issues as our client has ( for other reasons ) just -Xmx384m. 

## Details to the commits
### VW-001: Just test-debugging-code. Not meant to be in main-branch

#### What happens
the TextWidgetComponent.java has still a listener in KeyboardFocusManager.
#### When 
There are 2 Scenearios
##### Short sleep after close
When I do not sleep after closeDocument or really just a bit, then there is an Exception like 
```
Exception in thread "AWT-EventQueue-0" java.lang.NullPointerException: Cannot invoke "org.icepdf.ri.common.views.DocumentViewModel.getPageBoundary()" because "documentViewModel" is null
There are still some Listeners in memory
	at org.icepdf.ri.common.views.annotations.AbstractAnnotationComponent.<init>(AbstractAnnotationComponent.java:170)
	at org.icepdf.ri.common.views.annotations.acroform.TextWidgetComponent.<init>(TextWidgetComponent.java:52)
	at org.icepdf.ri.common.views.annotations.AnnotationComponentFactory.buildAnnotationComponent(AnnotationComponentFactory.java:133)
	at org.icepdf.ri.common.views.PageViewComponentImpl.initializeAnnotationsComponent(PageViewComponentImpl.java:568)

```
##### Bit longer sleep after close
When I increase the waiting time, then the propability of that memory-leak decreases, but there is no exception. I think it happens when the documentViewModel in DocumentViewControllerImpl.java is already null in Line 207. Maybe there is a way to prevent it like a closedListener but I did not find that..
<img width="1323" height="211" alt="image" src="https://github.com/user-attachments/assets/fbe98d21-f0ea-4b7a-bf3d-151c53385469" />
<img width="549" height="193" alt="image" src="https://github.com/user-attachments/assets/6fa22a11-1746-49d1-bb48-809a03ea1f71" />
### VW-002: FontManager with whitelistpattern when scanning local files
I tried to put all infos into the commit-msg
### VW-003: Option for different prefs-files
I tried to put all infos into the commit-msg
### VW-004: Cache the created Fonts.
At least in our case, this cache helps a lot.
"A picture is worth a thousand words."
#### Before
<img width="1023" height="1237" alt="image" src="https://github.com/user-attachments/assets/eb669fb6-5a18-45e6-9523-b176d7731ab1" />
#### After
<img width="1749" height="239" alt="image" src="https://github.com/user-attachments/assets/5f65b139-c272-41fb-8cde-d6fa8f2a3a64" />

## TODOS
If something of this PR would be considered to be part of the original main-branch, I would clean the added TODOS in the code, adding some documentation to the new System-Properties and add some unittests.

Commit:
VW-001: Just test-debugging-code. Not meant to be in main-branch

This is totally not production code. It was just for me to understand the problem, I had with the OOM in our client.

Commit:
VW-002: FontManager with whitelistpattern when scanning local files

Example to add a whitelist: System.getProperties().put(FILE_WHITE_LISTE_PATTERN_PROPERTY, MOST_COMMON_FONTS_PATTERN);

We had the first memory issue, when IcePdf was scanning local files. In my test-environment about 200 Fontfiles were scanned were some Unicode-Fonts had like 40Mbytes.

This commit give the option to add a pattern to filter the local fontfiles.

Commit:
VW-003: Option for different prefs-files

When FontManager would use a fileWhiteListePattern then it would make sence to save the result in a different prefs-file, if several apps with icepdf would run on that machine. A combination would be for example:

// adding a whitelist
System.getProperties().put(FILE_WHITE_LISTE_PATTERN_PROPERTY, MOST_COMMON_FONTS_PATTERN);
// new property to load/save the font-properties in a seperated property-file
System.getProperties().put(PREFERENCES_KEY_CLASS, MyApp.class);

Commit:
VW-004: Cache the created Fonts.

we had a document with a size of about 250kb and it needed many Mbyte in loaded in icepdf. The main reason was, that in memory existed about 50 FontFile-Object with basically the same content ( about 2 Mbyte each). This change reduced that to 1 instance

pcorless commented:
When you load documents in the small memory footprint, is it just one at a time or can more than one be open at a time?  Are the CPU's pretty quick?

VW-001 
This looks like a memory leak in the TextWidgetComponent class.  The sample file has a lot of TextWidget annotations.  Is this common PDF content for your workflow?  Do your users edit these fields?  If not the non-interactive view could be used to disable the widgets and free up some more memory, as the swing overlay overhead isn't insignificant.  

There is a DefaultPageViewLoadingListener for getting page loading events but there is nothing for document closing.  I'll take a closer look at this and see what I can find.  There is a case where a long running page parse/paint might cause a close issue but I don't think that's a issue here as the page content streams are pretty small for this type of content.

VW-002
The font scanning is supposed to be very light weight, opening and closing files and disposing of the stream once the properties are read.  I'll look into this as well as maybe something up with how I'm calling the fontbox api. This should be a very lightweight setup that only happens once per system. 

VW-003: 
I don't see a problem with allowing the font properties to be written to a different format/location.  That said I'd like to first investigate why VW-002 is so expense 

VW-004
This one is also interesting as there should be caching happening in the associated Document's library and ObjectLoader.  I can't quite see the root of the object tree in your screenshot.  I'll see if I can duplicated the issue with the associated sample file.  The widgets render might be the issue here again, but I'll need to poke around a bit more.   Again caching should be happening a lot lower in the loading process but I'll keep this caching suggestion in mind. 

I'll create tickets in the icepdf repo as I tackle each item.

volkerwahh commented:
Hi @pcorless ,
thx for the fast answer!
VW-002: I agree , that the font-scan is light weight. It must be some huge unicode fonts in our system...
Best Regards, 
Volker

volkerwahh commented:
Hi @pcorless ,
VW-002: 
Regarding the font-loading...at least at my mac the problem is pretty clear :-):
size in bytes of /System/Library/Fonts/Apple Color Emoji.ttc: 188589668 
But the user using our client in citrix-sessions in windows and don't know what happens if other huge files would come with an update or something else. 
For me the font-whitelist would be still helpful...
Best regards,
Volker

Commit:
VW-001: System.getProperties().put(...) is different as System.setProperty(..)

That was 0 with System.getProperties().put("...",3):
int cacheSize = Defs.intProperty(FONT_FACTORY_CACHE_SIZE_KEY, 0);

volkerwahh commented:
@pcorless 
> When you load documents in the small memory footprint, is it just one at a time or can more than one be open at a time? Are the CPU's pretty quick?

Damn, I was not answering your questions. I am getting old...yes...one by one and pretty quick cpu (Apple M1 Pro) :-) 

Btw. I changed a bit.  like a property to (de)activate the font-factory-caching, so it is easier to play around with the MemoryIssueViewerExample.java

thx4help & have a nice weekend!

pcorless commented:
When you render the PDF documents with acroform input fields,  are the user interacting with them at all, or are you just doing a static render of the PDF.  The reason I ask is if they aren't interacting then some memory would also be saved by not loading the swing components.

volkerwahh commented:
Hi Patrick,
I tried this: System.setProperty("org.icepdf.ri.viewer.readonly", "true");
Is that the "switch"? That gave me about 3 or 4 Mbyte. This is good, but does not help with our document(s)..
If I had to order it by "Priority":

- VW-004: most important as it is about >120Mbyte memory-usage
- VW-001: From time to time memory-Leak, that  TextWidgetComponent stays as Listener in KeyboardFocusManager. It happens sometimes when I run MemoryIssueViewerExample through that code: <img width="882" height="153" alt="image" src="https://github.com/user-attachments/assets/8afd3c0f-b509-4c01-a5ef-58b19b958158" />
- VW-002: We have also some big files in production, but not that big...So it is not that import 
- VW-003: I did that to use VW-002 more transparent. So also not that import

Regarding VW-004 btw a typical Stacktrace. "loadObject" has many different references (like the "111 0 R") which are basically all using one System-Font-URL in the FontFacotry:
<img width="1099" height="512" alt="image" src="https://github.com/user-attachments/assets/0888ae77-ba83-4b90-bde3-9035100c0a54" />

I hope this information is helping a bit..

Best regards, 
Volker

pcorless committed:
GH-439 add some protection and early exit of the PageImageCaptureTask if the view model is null and make sure the future is canceled on dispose.

pcorless commented:
The property for the annotation is org.icepdf.core.annotations.interactive.enabled=false.   I just pushed some code to fix a few widget components that didn't respect the property.  

Here are the two PR's for the vw-001 and vw-004.  Be great if you could look them over.  
https://github.com/pcorless/icepdf/pull/442
https://github.com/pcorless/icepdf/pull/443

Taking a few days off but will get back on the remaining two issue late next week.

Commit:
GH-439: Making methods dispose and initializeAnnotationsComponent in PageViewComponentImpl aware of each other in different threads

Commit:
VW-001: clean up MemoryIssueViewerExample

pcorless commented:
I've create two more PR to the remaining issues.  It would be great if you could review them.  With regards to https://github.com/pcorless/icepdf/pull/446 would it also make sense to add similar functionality to the ViewerPropertiesManager?

Pretty close to having enough for a 7.3.1 release.

volkerwahh commented:
Hi Patrick,
a 7.3.1 would be great! I'll have a look

volkerwahh commented:
Hi Patrick,
what do u think, when 7.3.1 could be released? 
Best Regards,
Volker

pcorless commented:
I did the release late last week.  I just double check maven central, 7.3.1 can be pulled.

volkerwahh commented:
arg, I could have checked it. Didn't think about it. 
The update of my feature-branch looks good and smooth and without memory-issues :)
